### PR TITLE
Added table storage deletion to the Storage layer and removed the del…

### PR DESCRIPTION
**Pull request by vikrambohra**
…etion logic from the catalog layer

## Summary
Added table location deletion logic in the Storage API layer. 
1. Created a Table deallocation method in the Storage interface for deleting table location along with implementations for various storages (LocalStorage, HdfsStorage, S3Storage)
2. Integrated the above logic into the catalog and repo layer. 

## Changes

- [ ] Client-facing API Changes
- [ ] Internal API Changes
- [ ] Bug Fixes
- [x] New Features
- [ ] Performance Improvements
- [ ] Code Style
- [x] Refactoring
- [ ] Documentation
- [ ] Tests

For all the boxes checked, please include additional details of the changes made in this pull request.  

## Testing Done

- [ ] Manually Tested on local docker setup. Please include commands ran, and their output.
- [ ] Added new tests for the changes made.
- [x] Updated existing tests to reflect the changes made.
- [ ] No tests added or updated. Please explain why. If unsure, please feel free to ask for help.
- [ ] Some other form of testing like staging or soak time in production. Please explain.

Updated LocalStorgeTest and BaseStorageTest with testing for deallocateTableLocation
# Additional Information
- [ ] Breaking Changes
- [ ] Deprecations
- [ ] Large PR broken into smaller PRs, and PR plan linked in the description.

For all the boxes checked, include additional details of the changes made in this pull request.

### CODE REVIEW

File: cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/BaseStorage.java [hunk 1]
```diff
@@ -76,4 +83,34 @@ public String allocateTableLocation(
...
+   * Deallocates/deletes Table Storage location Walk through the directory structure, sort the paths
+   * in reverse order to ensure files are deleted before directories, and deletes each path.
+   *
+   * @param location the base location of the table
+   * @param tableCreator the creator of the table
+   * @throws IOException
+   */
+  @Override
+  public void deallocateTableLocation(String location, String tableCreator) throws IOException {
+    List<IOException> exceptions = new ArrayList<>();
```
Comment (HotSushi) [inline cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/BaseStorage.java:97]: let's leave this as a no-op for `BaseStorage` and add some log for example: "deallocateTableLocation action not defined, skipping deallocation for Storage: XYZ"

Files.walk code will likely not work for storages outside of local system like gcs, or s3 or abs.

File: cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/local/LocalStorage.java [hunk 2]
```diff
@@ -65,4 +67,15 @@ public String allocateTableLocation(
...
   }
+  /**
+   * Deallocates/deletes Table Storage location recursively for Local storage
+   *
+   * @param location the base location of the table
+   * @param tableCreator the creator of the table
+   * @throws IOException
+   */
+  @Override
+  public void deallocateTableLocation(String location, String tableCreator) throws IOException {
```
Comment (HotSushi) [inline cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/local/LocalStorage.java:78]: 💯

File: cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/s3/S3Storage.java [hunk 3]
```diff
@@ -34,4 +36,22 @@ public StorageType.Type getType() {
...
+   * @param location the base location of the table
+   * @param tableCreator the creator of the table
+   * @throws IOException
+   */
+  @Override
+  public void deallocateTableLocation(String location, String tableCreator) throws IOException {
+    try {
+      this.s3StorageClient
+          .getNativeClient()
+          .deleteBucket(DeleteBucketRequest.builder().bucket(location).build());
```
Comment (HotSushi) [inline cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/s3/S3Storage.java:52]: currently bucket is defined at cluster level, this code will likely not work for bucket/db/tbl and it'll end up deleting all data.
can we check how S3FileIO implements  SupportsPrefixOperations? and use that code as reference

Comment (HotSushi) [reply cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/s3/S3Storage.java:52]: Lets also perform a docker compose test for this in s3

**Review by HotSushi — COMMENTED** [review]
PR is looking great, have a comment on BaseStorages implementation and s3 implementation. Also requesting a docker test for s3 minio setup.

File: services/tables/src/main/java/com/linkedin/openhouse/tables/repository/impl/OpenHouseInternalRepositoryImpl.java [hunk 4]
```diff
@@ -472,9 +473,25 @@ public boolean existsById(TableDtoPrimaryKey tableDtoPrimaryKey) {
...
         TableIdentifier.of(tableDtoPrimaryKey.getDatabaseId(), tableDtoPrimaryKey.getTableId()),
-        true);
+        false);
+
+    try {
+      storageManager.getDefaultStorage().deallocateTableLocation(location, tableCreator);
+    } catch (IOException e) {
+      String errMsg =
+          String.format(
+              "Exception while deleting table location %s for %s.%s",
```
Comment (maluchari) [inline services/tables/src/main/java/com/linkedin/openhouse/tables/repository/impl/OpenHouseInternalRepositoryImpl.java:491]: Consider logging the table location. We can skip location details in errors thrown back to clients to maintain table abstraction.

File: cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/Storage.java [hunk 5]
```diff
@@ -70,4 +71,13 @@ public interface Storage {
    */
   String allocateTableLocation(
       String databaseId, String tableId, String tableUUID, String tableCreator);
+
+  /**
+   * Deallocates/deletes Table Storage location
+   *
+   * @param location the base location of the table
+   * @param tableCreator the creator of the table
```
Comment (maluchari) [inline cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/Storage.java:79]: This need not tableCreator necessarily for all underlying storages.  May be actingPrincipal?

File: services/tables/src/main/java/com/linkedin/openhouse/tables/repository/impl/OpenHouseInternalRepositoryImpl.java [hunk 6]
```diff
@@ -472,9 +473,25 @@ public boolean existsById(TableDtoPrimaryKey tableDtoPrimaryKey) {
...
+    Table table =
+        catalog.loadTable(
+            TableIdentifier.of(
+                tableDtoPrimaryKey.getDatabaseId(), tableDtoPrimaryKey.getTableId()));
+    String location = table.location();
+    String tableCreator = table.properties().get("openhouse.tableCreator");
     catalog.dropTable(
         TableIdentifier.of(tableDtoPrimaryKey.getDatabaseId(), tableDtoPrimaryKey.getTableId()),
-        true);
+        false);
```
Comment (maluchari) [inline services/tables/src/main/java/com/linkedin/openhouse/tables/repository/impl/OpenHouseInternalRepositoryImpl.java:484]: Why do we still need this boolean? I think functionality wise this change will already impact any overridden implementations of the OpenHouseInternalCatalog until changes are made in the overridden implementations.  And also the boolean is used to determine whether this should be a soft or hard delete. Considering this logic is moved here should we still need the boolean flag?

File: cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/BaseStorage.java [hunk 7]
```diff
@@ -76,4 +83,34 @@ public String allocateTableLocation(
...
+  /**
+   * Deallocates/deletes Table Storage location Walk through the directory structure, sort the paths
+   * in reverse order to ensure files are deleted before directories, and deletes each path.
+   *
+   * @param location the base location of the table
+   * @param tableCreator the creator of the table
+   * @throws IOException
+   */
+  @Override
+  public void deallocateTableLocation(String location, String tableCreator) throws IOException {
```
Comment (sumedhsakdeo) [inline cluster/storage/src/main/java/com/linkedin/openhouse/cluster/storage/BaseStorage.java:96]: What is the main objective of this method? I understand what is does but why do we need it is not clear?
Can you update the PR description and javadoc?

**Review by sumedhsakdeo — CHANGES_REQUESTED** [review]

**Comment by vikrambohra** [comment]
Adding deallocateTableLocation in the storage interface does not make sense atm. The implementation in S3Storage is complicated and requires significant code duplication from S3FileIO.  Currently implemented internally in li-openhouse. Will revisit this if required

**Pull request closed without merging**